### PR TITLE
compaction: introduce multilevel write amp hueristic

### PR DESCRIPTION
**Pull request by msbutler**
This patch teaches pebble to pick a compaction involving 3 levels based on a
new WriteAmp heuristic, so long as the input level is below L0. Recall that
pebble#1603 introduced the mechanics to conduct a multilevel compaction by
teaching the compaction picker to call setupInputs() on the original output
level, finding files to include in a third level. The new WriteAmp heuristic
will choose this 3 level compaction if its predicted WriteAmp is less than the
original compaction's predicted WriteAmp. Multilevel compactions will be
considered for default and manual compactions, but not read triggered
compactions.

Predicted WriteAmp equals:
= sum(size of all input files)/sum(size of all files from higher levels)

According to several write intensive experiments, enabling the WriteAmp
heuristic has the potential to slightly reduce overall WriteAmplification,
increase throughput, and prevent inverted LSMs.
- See results from running pebble ycsb F with 64, 96, and 124 workers:
  https://github.com/cockroachdb/pebble/pull/2179#issuecomment-1353488791
- See write up from running the replay tool with kv0
  https://docs.google.com/document/d/1OmerofbtQfZUWQqLklqwmDJdVXI6qJzCLZQjsGirmFk/edit

Future work could tweak the heuristic or enable compactions with more than 3 levels.

Informs https://github.com/cockroachdb/pebble/issues/136

Release note: none

### CODE REVIEW

File: testdata/singledel_manual_compaction_set_with_del [hunk 1]
```diff
@@ -174,10 +174,8 @@ compact a-b L2
 ----
 1:
   000004:[a#10,SINGLEDEL-a#10,SINGLEDEL]
-3:
-  000008:[a#9,SET-a#8,DEL]
 4:
-  000007:[a#7,SET-a#7,SET]
+  000008:[a#9,SET-a#9,SET]
```
Comment (msbutler) [inline testdata/singledel_manual_compaction_set_with_del:178]: @jbowens while cleaning up the ML PR, I noticed the need change the results in this test case. Are ML compactions violating some invariant here?  Note that this test still fails on some later line.

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787)* status: 0 of 17 files reviewed, 5 unresolved discussions (waiting on @msbutler)

___
*[`event.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFYPY874S2hvOj-Ubq:-NbFYPY874S2hvOj-Ubr:bqx2yjo) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/event.go#L59)):*
> ```Go
> 		redact.Safe(formatFileNums(i.Tables)),
> 		redact.Safe(humanize.Bytes.Uint64(tablesTotalSize(i.Tables))),
> 		redact.Safe(i.Score))
> ```

I'm not sure if you're intending to keep this in when merged (I'm not opposed), but any updates to the compaction log entries require corresponding updates in the compaction log-parsing tool: https://github.com/cockroachdb/pebble/blob/master/tool/logs/compaction.go

I filed #2796 for the testing gap here.
___
*[`metrics.go` line 75 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFXCS0D4P7G-Zel9X6:-NbFXCS0D4P7G-Zel9X7:b-ab752f) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/metrics.go#L75)):*
> ```Go
> 	TablesMoved uint64
> 
> 	// BytesInTopML are the total bytes in a multilevel compaction coming from the top level.
> ```

nit: how about grouping within an anonymous struct, eg

```
MultiLevel struct {
    // BytesInTop ...
    BytesInTop uint64
    // BytesInTop ...
    BytesIn    uint64
    // BytesRead ...
    BytesRead  uint64
}
```
___
*[`metrics.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFXc713o3NS2Tt3QuY:-NbFXc713o3NS2Tt3QuZ:b-wcisye) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/metrics.go#L147)):*
> ```Go
> 		ReadCount         int64
> 		RewriteCount      int64
> 		SingleLevelCount  int64
> ```

can the single-level count be computed from the other counts?
___
*[`options.go` line 599 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFY0pC4kPFc52uysVC:-NbFY0pC4kPFc52uysVD:bi7k2x8) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/options.go#L599)):*
> ```Go
> 		MultiLevelCompactionHueristic MultiLevelHeuristic
> 
> 		MultiLevelCompactionAllowL0 bool
> ```

nit: What do you think of stashing this on the `WriteAmpHeuristic`/`MultiLevelHeuristic`?
___
*[`testdata/singledel_manual_compaction_set_with_del` line 178 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#gh-1284686226:-NbFVPy61MrwRKvYpDin:b-3d5tr8) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/testdata/singledel_manual_compaction_set_with_del#L178)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

@jbowens while cleaning up the ML PR, I noticed the need to change the results in this test case. Are ML compactions violating some invariant here?  Note that this test still fails on some later line.
</blockquote></details>

I don't think any invariants are being violated here, but this change from a L2->L3 compaction into a L2+L3->L4 multi-level compactions causes this test to no longer exercise the intended test case (a `DEL` and a `SINGLEDEL` meeting in a compaction). I think we should add a `define` flag that disables multi-level compactions and use it in this test (at line 147) keep the old picked compactions.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787)* status: 0 of 17 files reviewed, 5 unresolved discussions (waiting on @jbowens and @msbutler)

___
*[`event.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFYPY874S2hvOj-Ubq:-NbP8h_15nGR39X8SETH:b-dfhi33) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/event.go#L59)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I'm not sure if you're intending to keep this in when merged (I'm not opposed), but any updates to the compaction log entries require corresponding updates in the compaction log-parsing tool: https://github.com/cockroachdb/pebble/blob/master/tool/logs/compaction.go

I filed #2796 for the testing gap here.
</blockquote></details>

Ah good point. Before I go make those changes, do you have any suggestions for how I've modified the compaction log lines?

A default compaction now displays the compaction score of each input level and the single and ML overlapping ratio. As an example, here's how a single level compaction's logs line currently look in my PR:
```
[JOB 92] compacting(default) L4 [250870] (869KB) Score=1.01 + L5 [250859] (6.8MB) Score=0.99; OverlappingRatio: Single 8.03, Multi 25.05;
[JOB 92] compacted(default) L4 [250870] (869KB) Score=1.01 + L5 [250859] (6.8MB) Score=0.99 -> L5 [250871] (7.1MB), in 0.1s (0.1s total), output rate 111MB/s
```

And a multi level:

```
[JOB 11] compacting(default) [multilevel] L2 [250858] (2.1MB) Score=1.09 + L3 [247985 247989 247848] (17MB) Score=0.99 + L4 [250817 250834 238396] (28MB) Score=1.00; OverlappingRatio: Single 3.77, Multi 1.46;
[JOB 11] compacted(default) [multilevel] L2 [250858] (2.1MB) Score=1.09 + L3 [247985 247989 247848] (17MB) Score=0.99 + L4 [250817 250834 238396] (28MB) Score=1.00 -> L4 [250859 250860 250861 250862 250863] (46MB), in 0.2s (0.2s total), output rate 185MB/s
```
___
*[`metrics.go` line 75 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFXCS0D4P7G-Zel9X6:-NbLLQOQ-B_T4NqwklxF:b-kghtrs) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/metrics.go#L75)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: how about grouping within an anonymous struct, eg

```
MultiLevel struct {
    // BytesInTop ...
    BytesInTop uint64
    // BytesInTop ...
    BytesIn    uint64
    // BytesRead ...
    BytesRead  uint64
}
```
</blockquote></details>

good point. done.
___
*[`metrics.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFXc713o3NS2Tt3QuY:-NbLM2f90NFA3Y5yMQ3n:b63dbho) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/metrics.go#L147)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

can the single-level count be computed from the other counts?
</blockquote></details>

Deleted. Wasn't really necessary.
___
*[`options.go` line 599 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFY0pC4kPFc52uysVC:-NbLNzqq4Zx3hdXr-PAI:bfqzzl3) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/options.go#L599)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: What do you think of stashing this on the `WriteAmpHeuristic`/`MultiLevelHeuristic`?
</blockquote></details>

Good nit. Fixed.
___
*[`testdata/singledel_manual_compaction_set_with_del` line 178 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#gh-1284686226:-NbLL5DV4XFRdC7UyzR-:b-r2hte8) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/testdata/singledel_manual_compaction_set_with_del#L178)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I don't think any invariants are being violated here, but this change from a L2->L3 compaction into a L2+L3->L4 multi-level compactions causes this test to no longer exercise the intended test case (a `DEL` and a `SINGLEDEL` meeting in a compaction). I think we should add a `define` flag that disables multi-level compactions and use it in this test (at line 147) keep the old picked compactions.
</blockquote></details>

Sounds good. Added that define flag.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787)* status: 0 of 19 files reviewed, 3 unresolved discussions (waiting on @jbowens and @msbutler)

___
*[`event.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFYPY874S2hvOj-Ubq:-NbQU--kBLQFwZUI7xS8:b-ctmcr7) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/event.go#L59)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Ah good point. Before I go make those changes, do you have any suggestions for how I've modified the compaction log lines?

A default compaction now displays the compaction score of each input level and the single and ML overlapping ratio. As an example, here's how a single level compaction's logs line currently look in my PR:
```
[JOB 92] compacting(default) L4 [250870] (869KB) Score=1.01 + L5 [250859] (6.8MB) Score=0.99; OverlappingRatio: Single 8.03, Multi 25.05;
[JOB 92] compacted(default) L4 [250870] (869KB) Score=1.01 + L5 [250859] (6.8MB) Score=0.99 -> L5 [250871] (7.1MB), in 0.1s (0.1s total), output rate 111MB/s
```

And a multi level:

```
[JOB 11] compacting(default) [multilevel] L2 [250858] (2.1MB) Score=1.09 + L3 [247985 247989 247848] (17MB) Score=0.99 + L4 [250817 250834 238396] (28MB) Score=1.00; OverlappingRatio: Single 3.77, Multi 1.46;
[JOB 11] compacted(default) [multilevel] L2 [250858] (2.1MB) Score=1.09 + L3 [247985 247989 247848] (17MB) Score=0.99 + L4 [250817 250834 238396] (28MB) Score=1.00 -> L4 [250859 250860 250861 250862 250863] (46MB), in 0.2s (0.2s total), output rate 185MB/s
```
</blockquote></details>

Perhaps its also worth bikeshedding how we'd want to modify the window logger. I think the current impl looks like this: https://github.com/cockroachdb/pebble/commit/b5deee5d78f3a688338359265e557d90ed3cf77b 


One nice quality of this format is that it cleanly partitions the kinds of compactions. While ML currently can only get triggered in a default compaction, we could enable it for move and read triggered compactions in the future if we wanted to. I propose adding an `extra(B)` column between the `in(B)` and `out(B)` columns and keep everything else the same.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787)* status: 0 of 19 files reviewed, 3 unresolved discussions (waiting on @jbowens and @msbutler)

___
*[`event.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFYPY874S2hvOj-Ubq:-NbRBaoZ3c20nW5OuZ8o:b-bc18s6) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/event.go#L59)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

Perhaps its also worth bikeshedding how we'd want to modify the window logger. I think the current impl looks like this: https://github.com/cockroachdb/pebble/commit/b5deee5d78f3a688338359265e557d90ed3cf77b 


One nice quality of this format is that it cleanly partitions the kinds of compactions. While ML currently can only get triggered in a default compaction, we could enable it for move and read triggered compactions in the future if we wanted to. I propose adding an `extra(B)` column between the `in(B)` and `out(B)` columns and keep everything else the same. 
</blockquote></details>

^well actually, with the given window format, the in(B) column represents the sum of bytes over all input files and out(B) represents the size of the new file created from the compaction. Given this, it wouldn't make much sense to just log the bytes contributed from the extra level. 

We could split in(B) into 3 columns (top level, extra level, output level), but that seems out of scope in this PR.

For this PR, is it alright if I just update the tests and ensure ML compaction log lines get parsed correctly?

**Review by jbowens — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787)* status: 0 of 20 files reviewed, 1 unresolved discussion (waiting on @msbutler)

___
*[`event.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFYPY874S2hvOj-Ubq:-NbUuiQUAvPwKYfaDiY9:b-t7yzbm) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/event.go#L59)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

^well actually, with the given window format, the in(B) column represents the sum of bytes over all input files and out(B) represents the size of the new file created from the compaction. Given this, it wouldn't make much sense to just log the bytes contributed from the extra level. 

We could split in(B) into 3 columns (top level, extra level, output level), but that seems out of scope in this PR.

For this PR, is it alright if I just update the tests and ensure ML compaction log lines get parsed correctly?

</blockquote></details>

Yeah, I think that's alright. Mind creating an issue for the follow up though?

**Comment by msbutler** [comment]
@sumeerbhola @jbowens RFAL! there were 2 minor test failures bc I didn't update the test case log lines correctly. I will update these after y'all have a read.

File: tool/logs/testdata/compactions [hunk 2]
```diff
@@ -120,7 +120,7 @@ reset
```
Comment (msbutler) [inline tool/logs/testdata/compactions:120]: I will plan to update the logs in this test file if y'all are cool with my log line updates.

**Review by msbutler — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787)* status: 9 of 25 files reviewed, 4 unresolved discussions (waiting on @jbowens and @sumeerbhola)

___
*[`compaction_picker.go` line 1407 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2787#-Nd7U1Bx6R_lj-Z-iO78:-NdGhShZ1vGA1klT3WuF:bcvjgzp) ([raw file](https://github.com/cockroachdb/pebble/blob/deeb6ed8e7c09f18e38307b69bcff96e06583f59/compaction_picker.go#L1407)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

`pickedCompaction` is a hefty struct. Can we take this by pointer?

And in that case, should we even bother returning the value out too? Call sites are modifying the picked compaction that they pass in.
</blockquote></details>

Good idea. Done.
___
*[`compaction_picker.go` line 1412 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2787#-Nd7U8PyBWERHswK834A:-NdGhddbA6exuTzxQTh7:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/deeb6ed8e7c09f18e38307b69bcff96e06583f59/compaction_picker.go#L1412)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: s/7/numLevels/
</blockquote></details>

Done.
___
*[`compaction_picker.go` line 1763 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NdAvu4F7qZW9Eap9S_U:-NdGiYU42JRpY3kphGlt:bfluihm) ([raw file](https://github.com/cockroachdb/pebble/blob/deeb6ed8e7c09f18e38307b69bcff96e06583f59/compaction_picker.go#L1763)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I'm mildly concerned about the amount of extra work involved in cloning and re-constructing inputs. Compaction-picking holds d.mu, which prevents WAL rotations, in-progress flushes and compactions from completing, etc. How about leaving a TODO to microbenchmark and possibly restructure to limit duplicate work here?
</blockquote></details>

Added the todo
___
*[`event.go` line 59 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2787#-NbFYPY874S2hvOj-Ubq:-NdGil7I-OLzV8sw3C6c:b-hzjo8m) ([raw file](https://github.com/cockroachdb/pebble/blob/773965ee79d664e88a7d7da6f75c8b1c8569e9cc/event.go#L59)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Yeah, I think that's alright. Mind creating an issue for the follow up though?
</blockquote></details>

Done. https://github.com/cockroachdb/pebble/issues/2814
___
*[`options.go` line 1077 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2787#-Nd7V_-20PFyiaNli6-x:-NdGizID-vIQ02ZW-sSL:b-3r619u) ([raw file](https://github.com/cockroachdb/pebble/blob/deeb6ed8e7c09f18e38307b69bcff96e06583f59/options.go#L1077)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: I'm noticing now that heuristic is misspelled here.
</blockquote></details>

heh. fixed. 
___
*[`tool/logs/testdata/compactions` line 120 at r3](https://reviewable.io/reviews/cockroachdb/pebble/2787#gh-1291376293:-NdGrHPy5fxvmEl2iYkd:b-zgnxi6) ([raw file](https://github.com/cockroachdb/pebble/blob/deeb6ed8e7c09f18e38307b69bcff96e06583f59/tool/logs/testdata/compactions#L120)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

I will plan to update the logs in this test file if y'all are cool with my log line updates.
</blockquote></details>

Just Updated.

**Review by jbowens — APPROVED** [review]
:lgtm_strong: 

> Reviewed 1 of 7 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2787#-:-Ndf7rsy8NLOasvTsMEG:b-uwkowd)* status: 10 of 25 files reviewed, 1 unresolved discussion (waiting on @sumeerbhola)

**Comment by msbutler** [comment]
rebased to address a few merge conflicts. nothing complicated.

**Pull request merged**